### ISSUE TITLE
Turn pristine back to true if model back to original attributes

### ISSUE THREAD

**Issue by yamsellem** [davidkpiano/react-redux-form#685]
A `pristine` flag indicates that a model has not been touched by the user. But, this flag is not turned back to `true` (meaning, nothing has changed) when the user change the value back to its original value.

Is there a way (or another flag) to know that a model has not changed since a previous state?

What I'm trying to achieve is to know if a model has changed between two actions that save it to the server. If, after a save it has not changed, don't save it again.

Thanks.

**Comment by davidkpiano** [comment]
> What I'm trying to achieve is to know if a model has changed between two actions that save it to the server. If, after a save it has not changed, don't save it again.

This is use-case specific - we can't make blanket assumptions here because some APIs might change the server-side value externally, so even though it didn't change client-side, it _did_ change server-side.

All field values have an `.initialValue` prop that you can compare to. You can `connect()` to get that field value from the store (e.g., `forms.user.name.initialValue`) and manually compare.

Other than that, what does an ideal API look like to you for this use-case (with regard to not making assumptions about the server)?

**Comment by yamsellem** [comment]
@davidkpiano you're right, the server can change values when it answers.
I understand with your words that this is not possible — yet — with react redux form. Right?

Something useful could be `formActions.change('forms.model, model, {sync: true})` triggered manually by the action after a successful http call. This is exactly like `formActions.change('forms.model ,model)` but it resets `pristine` back to true.

But this is half of the problem.
When the user update a value, and update it back to its original value, I guess `.initialValue` will do the trick.

Sounds great?

**Comment by davidkpiano** [comment]
I mean, I could add `.isInitial`:

```js
// initial user: { name: 'david' }

dispatch(actions.change('user.name', 'john'));

forms.user.name.isInitial; // false

dispatch(actions.change('user.name', 'david'));

forms.user.name.isInitial; // true
```

If this gets enough interest (vote with 👍 or 👎 ), I'll add it!

**Comment by yamsellem** [comment]
@davidkpiano how this `.isInitial` will be reseted to true on server sync?

**Comment by davidkpiano** [comment]
> how this .isInitial will be reseted to true on server sync?

That's, again, use-case specific, depending on how you're syncing with the server already in your client-side code.

If the server changes, you can always `load` the new value:

```js
API.getUpdatedValue().then(newValue => {
  dispatch(actions.load('user.name', newValue));
});
```

Then `.value` will be compared to the `.loadedValue`.

**Comment by yamsellem** [comment]
@davidkpiano dispatching `load` — `change` silently — will not rerender components.
So, on a server sync, `load` cannot be used, I guess.

IMO, syncing a form with a server and knowing it's in sync it's not really use case specific. Syncing with a server is a form fate ;-).

If this isInitial can also be manually reset it's perfect, like `formActions.change('forms.model', model, {initial: true})`.

**Comment by yamsellem** [comment]
@davidkpiano any news about this? Thanks ;-).

**Comment by davidkpiano** [comment]
Still deciding on the API. The problem with `.initial` is that it's very close to `.initialValue`, which can cause some confusion. Also, `.load` _will_ rerender control components, have you tried that?

**Comment by yamsellem** [comment]
@davidkpiano maybe `.unchanged` should do the trick?

Using `load` as you said don't solve my issue, because, I'm also trying to activate / deactivate back and save buttons. An `.unchanged` attribute and its `formActions.change('forms.model', model, {unchanged: true})` will greatly help ;-).

_ps. by the way, I don't clearly see the point of `pristine`. This `.unchanged` sound way useful to me._

**Comment by davidkpiano** [comment]
I like `.unchanged`.

The difference between `.pristine` and `.unchanged` is that `.pristine` signifies if a control has been changed at all, even if it was changed to its original value. It's more about the life cycle of the control than its actual value, which is useful for displaying validation.

If you can send me a minimal code example of what you're trying to do, I might be able to help better. My thinking is that this is the exact use-case for `.load`, so there must be some way for that to work with your use-case.

**Comment by yamsellem** [comment]
@davidkpiano if I understand correctly what you said, `formActions.load('forms.model', model)` resets initialValues and pristine? And `formActions.change('forms.model', model)` updates only the model and sets `pristine` to `false` and don't update `initialValues`?

My use case:
* fetch data from the server
* display a form that the user can interact with
* render `back` and `save` buttons according to the `unchanged` state of the model

**Comment by yamsellem** [comment]
@davidkpiano so, where do we go from here?

**Comment by davidkpiano** [comment]
The behavior of `.load` just does a silent `.change`, and when a field is reset, it is reset to the `.loadedValue` instead of the `.initialValue`. It does not change the `.initialValue`, nor odes it change `.pristine` (because it's a silent change).

I think the simplest change, for now, would be to add an `.unchanged` prop to the field state. Would that work for your use-case? Here's the logic:

- if `.loadedValue !== undefined && .value === .loadedValue`, then `.unchanged = true`
- else, if `.value === .initialValue` then `.unchanged = true`
- else, `.unchanged = false`

**Comment by yamsellem** [comment]
Yes, that behavior seems perfect to me.

I am wondering what will be the impact of silencing every server sync — even the first one?

**Comment by davidkpiano** [comment]
> I am wondering what will be the impact of silencing every server sync — even the first one?

What do you mean by "silencing"?

**Comment by yamsellem** [comment]
By using `load` or the `silent` option.

**Comment by yamsellem** [comment]
@davidkpiano _to rephrase it_: what is the impact of silencing every server sync with `formActions.load('forms.model', model)` — even the first one? 

The scenario could be the following:
* an api call that `load` the model (`GET /model`)
* a user input that `change` the model
* a user save that triggers an api call that `load` the model (`PUT /model`)
* a user input that `change` the model
* ...

The goal is to properly display:
* a `save` button only if the model is not `.unchanged` since the last sync (aka the last `load`)
* a `cancel` button that display a `are your sure you want to ignore your updates?` only if the model is not `.unchanged` since the last sync (aka the last `load`)

Something is not quite clear in the doc: "`actions.load(model, value)` is equivalent to `actions.change(model, value, { silent: true })`". But IMO `load`is also updating `.loadedValue` which `change` does not. Am I right?

**Comment by davidkpiano** [comment]
> "actions.load(model, value) is equivalent to actions.change(model, value, { silent: true })"

You're right, there is a subtle difference. Can you open a new documentation ticket about that (and about `.loadedValue` for tracking?
